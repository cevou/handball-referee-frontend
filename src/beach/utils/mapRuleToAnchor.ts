export default function mapRuleToAnchor(rule: string, language: string) {
  const matchBasicRule = rule.match(/^([0-9]+:[0-9]+)/);

  if (matchBasicRule) {
    return `#${matchBasicRule[1]}`;
  }

  const mapClarification = rule.match(/^Clarification\sNo.\s([0-9]+)/);
  if (mapClarification) {
    switch (mapClarification[1]) {
      case "1":
        switch (language) {
          default:
            return "#1.-awarding-of-points";
        }
      case "2":
        switch (language) {
          default:
            return "#2.-special-circumstances";
        }
      case "3":
        switch (language) {
          default:
            return "#3.-free-throw-execution-after-the-final-signal";
        }
      case "4":
        switch (language) {
          default:
            return "#4.-passive-play";
        }
      case "5":
        switch (language) {
          default:
            return "#5.-unsportsmanlike-conduct";
        }
      case "6":
        switch (language) {
          default:
            return "#6.-seriously-unsportsmanlike-conduct";
        }
      case "7":
        switch (language) {
          default:
            return "#7.-definition-of-a-clear-chance-of-scoring‚Äù";
        }
      case "8":
        switch (language) {
          default:
            return "#8.-interruption-by-the-timekeeper";
        }
      case "9":
        switch (language) {
          default:
            return "#9.-shoot-out-and-fast-break";
        }
      default:
        switch (language) {
          default:
            return "#clarifications-to-the-rules-of-the-game";
        }
    }
  }

  const matchSAR = rule.match(/^Substitution Area Regulations/);
  if (matchSAR) {
    switch (language) {
      default:
        return "#substitution-area-regulations";
    }
  }

  const matchHandSignals = rule.match(/^Hand signals/);
  if (matchHandSignals) {
    switch (language) {
      default:
        return "#hand-signals";
    }
  }

  return "";
}
