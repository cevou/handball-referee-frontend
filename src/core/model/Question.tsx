/* eslint-disable no-underscore-dangle */
import { IDBPDatabase } from "idb";
import {
  IAnswer, IQuestion, ITestData, RefereeDB,
} from "./index";

export default class Question {
  private _id: string;

  private _rule: string;

  private number: number|null;

  private _question: { [lang: string]: string };

  private _answers: { [lang: string]: { [key: string]: string } };

  private _correct: string[];

  private _rules: string[];

  private _box = 1;

  private lastAsked: Date | null = null;

  private previousResults: boolean[] = [];

  private _numAsked = 0;

  private _numCorrect = 0;

  private _numWrong = 0;

  constructor(lang: string, question: IQuestion, answer: IAnswer, testData?: ITestData) {
    const [rule, number] = question.id.split(".");

    this._id = question.id;
    this._rule = rule;
    this.number = number ? parseInt(number, 10) : null;
    this._question = {
      [lang]: question.question,
    };
    this._answers = {
      [lang]: question.answers,
    };
    this._correct = answer.correct;
    this._rules = answer.rule;

    if (testData) {
      this._box = testData.box;
      this.lastAsked = testData.lastAsked;
      this.previousResults = testData.previousResults;
      this._numAsked = testData.asked;
      this._numCorrect = testData.correct;
      this._numWrong = testData.wrong;
    }
  }

  get id(): string {
    return this._id;
  }

  get rule(): string {
    return this._rule;
  }

  get question(): { [p: string]: string } {
    return this._question;
  }

  get answers(): { [p: string]: { [p: string]: string } } {
    return this._answers;
  }

  get box(): number {
    return this._box;
  }

  get correct(): string[] {
    return this._correct;
  }

  get rules(): string[] {
    return this._rules;
  }

  get numWrong(): number {
    return this._numWrong;
  }

  get numCorrect(): number {
    return this._numCorrect;
  }

  get numAsked(): number {
    return this._numAsked;
  }

  public updateData(lang: string, question: IQuestion) {
    this._question[lang] = question.question;
    this._answers[lang] = question.answers;
  }

  public checkAnswer(answers: string[]) {
    let answeredCorrect = true;
    if (answers.length !== this._correct.length) {
      answeredCorrect = false;
    } else {
      for (let i = 0; i < answers.length; i += 1) {
        if (answers[i] !== this._correct[i]) {
          answeredCorrect = false;
          break;
        }
      }
    }

    this.lastAsked = new Date();
    this._numAsked += 1;
    this.previousResults.splice(0, 0, answeredCorrect);
    if (answeredCorrect) {
      this._numCorrect += 1;
      this._box = Math.min(5, this._box + 1);
    } else {
      this._numWrong += 1;
      this._box = 1;
    }

    return {
      answeredCorrect,
      correct: this._correct,
      rules: this._rules,
    };
  }

  public async persist(db: IDBPDatabase<RefereeDB>) {
    await db.put("questions", {
      box: this._box,
      asked: this._numAsked,
      correct: this._numCorrect,
      wrong: this._numWrong,
      previousResults: this.previousResults,
      lastAsked: this.lastAsked,
    }, this.id);
  }
}
