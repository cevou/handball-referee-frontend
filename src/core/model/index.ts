import { DBSchema } from "idb";

export interface IQuestion {
  id: string;
  question: string;
  answers: { [key: string]: string };
}

export interface IAnswer {
  id: string;
  correct: string[];
  rule: string[];
}

export interface ITestData {
  box: number;
  lastAsked: Date | null;
  previousResults: boolean[];
  asked: number;
  correct: number;
  wrong: number;
}

export interface ITestResponse {
  correct: string[];
  rules: string[];
  answeredCorrect?: boolean;
}

export interface RefereeDB extends DBSchema {
  questions: {
    key: string;
    value: ITestData;
    indexes: {
      pick: [number, Date];
    };
  };
}
