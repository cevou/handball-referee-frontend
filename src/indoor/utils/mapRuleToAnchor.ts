export default function mapRuleToAnchor(rule: string, language: string) {
  const matchBasicRule = rule.match(/^([0-9]+:[0-9]+)/);

  if (matchBasicRule) {
    return `#${matchBasicRule[1]}`;
  }

  const mapClarification = rule.match(/^Clarification\s([0-9]+)/);
  if (mapClarification) {
    switch (mapClarification[1]) {
      case "2":
        switch (language) {
          case "de":
            return "#2.-time-out";
          case "es":
            return "#2.-tiempo-muerto";
          default:
            return "#2.-time-out";
        }
      case "3":
        switch (language) {
          case "de":
            return "#3.-team-time-out";
          case "es":
            return "#3.-tiempo-muerto-de-equipo";
          default:
            return "#3.-team-time-out";
        }
      case "4":
        switch (language) {
          case "de":
            return "#4.-passives-spiel";
          case "es":
            return "#4.-juego-pasivo";
          default:
            return "#4.-passive-play";
        }
      case "6":
        switch (language) {
          case "de":
            return "#6.-definition-einer-klaren-torgelegenheit";
          case "es":
            return "#6.--definición-de-clara-ocasión-de-gol";
          default:
            return "#6.-definition-of-clear-chance-of-scoring";
        }
      case "7":
        switch (language) {
          case "de":
            return "#7.-eingreifen-durch-den-zeitnehmer";
          case "es":
            return "#7.-intervención/interrupción-por-parte-del-cronometrador-o-del-delegado";
          default:
            return "#7.-intervention-by-the-timekeeper-or-a-delegate";
        }
      case "8":
        switch (language) {
          case "de":
            return "#8.-verletzter-spieler";
          case "es":
            return "#8.-jugador-lesionado";
          default:
            return "#8.-injured-player";
        }
      default:
        switch (language) {
          case "de":
            return "#erläuterungen-zu-den-spielregeln";
          case "es":
            return "#aclaraciones-a-las-reglas-de-juego";
          default:
            return "#clarifications-to-the-rules-of-the-game";
        }
    }
  }

  const mapGuidelines = rule.match(/^Guidelines(?:\s(.+))?|^New Guideline/);
  if (mapGuidelines) {
    switch (mapGuidelines[1]) {
      case "8:10c":
        switch (language) {
          case "de":
            return "#abstandsvergehen";
          case "es":
            return "#no-respetar-la-distancia";
          default:
            return "#not-respecting-the-distance";
        }
      case "8:10d":
        switch (language) {
          case "de":
            return "#disqualifikation-in-den-letzten-30-sekunden";
          case "es":
            return "#descalificación-durante-los-últimos-30-segundos-de-juego";
          default:
            return "#disqualification-during-the-last-30-seconds";
        }
      default:
        switch (language) {
          case "de":
            return "#guidelines-und-interpretationen";
          case "es":
            return "#directrices-e-interpretaciones";
          default:
            return "#guidelines-and-interpretations";
        }
    }
  }

  const matchSAR = rule.match(/^Substitution Area Regulations/);
  if (matchSAR) {
    switch (language) {
      case "de":
        return "#auswechselraum-reglement";
      case "es":
        return "#reglamento-relativo-a-la-zona-de-cambios";
      default:
        return "#substitution-area-regulations";
    }
  }

  const matchHandSignals = rule.match(/^Hand signals/);
  if (matchHandSignals) {
    switch (language) {
      case "de":
        return "#handzeichen";
      case "es":
        return "#gestoforma";
      default:
        return "#hand-signals";
    }
  }

  const matchEquipment = rule.match(/^Regulations on Protective Equipment and Accessories/);
  if (matchEquipment) {
    switch (language) {
      case "de":
        return "#ausrüstungsreglement";
      case "es":
        return "#reglamento-sobre-equipos-de-protección-y-accesorios";
      default:
        return "#regulations-on-protective-equipment-and-accessories";
    }
  }

  return "";
}
