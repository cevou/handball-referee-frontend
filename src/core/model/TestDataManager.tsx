/* eslint-disable no-underscore-dangle */
import { IDBPDatabase, openDB } from "idb";
import {
  IAnswer, IQuestion, ITestData, RefereeDB,
} from "./index";
import Question from "./Question";

export default class TestDataManager {
  private db: IDBPDatabase<RefereeDB> | null = null;

  private _data: { [id: string]: Question } = {};

  private loadedLanguages: string[] = [];

  private todo: string[] = [];

  private initialized = false;

  private currentId = "";

  private _asked = 0;

  private _correct = 0;

  private _wrong = 0;

  private _answerData: Array<IAnswer> = [];

  constructor(answerData: Array<IAnswer>) {
    this._answerData = answerData;
  }

  public async initialize(language: string) {
    if (this.initialized) {
      await this.switchLanguage(language);
      return this._data[this.currentId];
    }

    const mappedAnswers = this._answerData.reduce<{ [id: string]: IAnswer }>((prev, curr) => ({
      ...prev,
      [curr.id]: curr,
    }), {});

    const questions = await TestDataManager.loadQuestions(language);

    this.loadedLanguages = [language];

    const mappedQuestions = questions.reduce<{ [id: string]: IQuestion }>((prev, curr) => ({
      ...prev,
      [curr.id]: curr,
    }), {});

    const mappedTestData: { [id: string]: ITestData } = {};

    let db;

    // idb library does not support IE
    const ua = window.navigator.userAgent;
    const isIE = /MSIE|Trident/.test(ua);
    if (!isIE) {
      db = await openDB<RefereeDB>("referee", 1, {
        async upgrade(currentDB, oldVersion) {
          if (oldVersion < 1) {
            currentDB.createObjectStore("questions");
          }
        },
      });
    }

    const ids = Object.keys(mappedAnswers);
    this.todo = [...ids];

    if (db) {
      const tx = db.transaction("questions", "readwrite");

      const currentDate = new Date();

      let cursor = await tx.store.openCursor();

      while (cursor) {
        mappedTestData[cursor.key] = cursor.value;

        const testData = cursor.value;

        this._asked += testData.asked;
        this._correct += testData.correct;
        this._wrong += testData.wrong;

        if (testData.lastAsked) {
          let remove = false;
          const diffTime = Math.abs(+testData.lastAsked - +currentDate);
          const diffDays = Math.floor(diffTime / (1000 * 60 * 60 * 24));

          // remove if needed
          if (
            (testData.box === 2 && diffDays < 1)
            || (testData.box === 3 && diffDays < 3)
            || (testData.box === 4 && diffDays < 7)
            || (testData.box === 5 && diffDays < 30)
          ) {
            remove = true;
          }

          if (remove) {
            // eslint-disable-next-line no-loop-func
            const index = this.todo.findIndex((val) => val === cursor?.key);
            if (index > -1) {
              this.todo.splice(index, 1);
            }
          }
        }

        // eslint-disable-next-line no-await-in-loop
        cursor = await cursor.continue();
      }

      await tx.done;

      this.db = db;
    }

    for (let i = 0; i < ids.length; i += 1) {
      const id = ids[i];
      this._data[id] = new Question(
        language,
        mappedQuestions[id],
        mappedAnswers[id],
        mappedTestData[id],
      );
    }

    this.initialized = true;

    return this.next();
  }

  get wrong(): number {
    return this._wrong;
  }

  get correct(): number {
    return this._correct;
  }

  get asked(): number {
    return this._asked;
  }

  get data(): { [p: string]: Question } {
    return this._data;
  }

  public async checkAnswer(answers: string[]) {
    if (!this.currentId) {
      return {
        correct: [],
        rules: [],
      };
    }

    // persist test data

    const question = this._data[this.currentId];

    const result = question.checkAnswer(answers);

    this._asked += 1;
    if (result.answeredCorrect) {
      this._correct += 1;
    } else {
      this._wrong += 1;
    }

    if (this.db) {
      await question.persist(this.db);
    }

    if (question.box > 1) {
      const index = this.todo.findIndex((val) => val === question.id);
      this.todo.splice(index, 1);
    }

    return result;
  }

  public next() {
    this.currentId = this.todo[Math.floor(Math.random() * this.todo.length)];
    return this._data[this.currentId];
  }

  private async switchLanguage(language: string) {
    if (this.loadedLanguages.includes(language)) {
      return;
    }

    const questions = await TestDataManager.loadQuestions(language);
    for (let i = 0; i < questions.length; i += 1) {
      const question = questions[i];
      this._data[question.id].updateData(language, question);
    }

    this.loadedLanguages = [...this.loadedLanguages, language];
  }

  private static async loadQuestions(language: string): Promise<IQuestion[]> {
    const response = await fetch(`/data/questions/${language}.json`);
    const json = await response.json();
    if (!response.ok) {
      throw new Error(json);
    }

    return json;
  }
}
