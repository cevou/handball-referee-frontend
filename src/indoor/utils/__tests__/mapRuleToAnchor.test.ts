import { IAnswer } from "../../../core/model";
import mapRuleToAnchor from "../mapRuleToAnchor";

// eslint-disable-next-line @typescript-eslint/no-var-requires
const Answers = require("../../data/answers.json");

describe("mapRuleToAnchor", () => {
  it("should map all available rules to the correct anchor", () => {
    const texts: string[] = Array.from(
      new Set(Answers.reduce((all: string[], answer: IAnswer) => all.concat(answer.rule), [])),
    );
    texts.sort();

    const en = texts.map((text) => mapRuleToAnchor(text, "en"));
    const de = texts.map((text) => mapRuleToAnchor(text, "de"));
    const es = texts.map((text) => mapRuleToAnchor(text, "es"));

    expect(en).toMatchSnapshot();
    expect(de).toMatchSnapshot();
    expect(es).toMatchSnapshot();
  });
});
