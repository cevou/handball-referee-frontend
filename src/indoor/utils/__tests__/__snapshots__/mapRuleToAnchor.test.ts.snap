// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mapRuleToAnchor should map all available rules to the correct anchor 1`] = `
Array [
  "#10:1",
  "#10:2",
  "#10:3",
  "#10:4",
  "#11:1",
  "#11:4",
  "#11:5",
  "#12:1",
  "#12:1",
  "#12:2",
  "#13:1",
  "#13:1",
  "#13:1",
  "#13:2",
  "#13:3",
  "#13:4",
  "#13:4",
  "#13:4",
  "#13:5",
  "#13:6",
  "#13:6",
  "#13:7",
  "#13:8",
  "#14:1",
  "#14:10",
  "#14:1",
  "#14:1",
  "#14:1",
  "#14:2",
  "#14:3",
  "#14:4",
  "#14:5",
  "#14:6",
  "#14:7",
  "#14:8",
  "#14:9",
  "#15:1",
  "#15:2",
  "#15:3",
  "#15:4",
  "#15:5",
  "#15:5",
  "#15:5",
  "#15:6",
  "#15:7",
  "#15:7",
  "#15:7",
  "#15:8",
  "#15:9",
  "#16:1",
  "#16:10",
  "#16:11",
  "#16:11",
  "#16:1",
  "#16:1",
  "#16:3",
  "#16:3",
  "#16:3",
  "#16:3",
  "#16:3",
  "#16:3",
  "#16:3",
  "#16:3",
  "#16:4",
  "#16:5",
  "#16:6",
  "#16:6",
  "#16:6",
  "#16:6",
  "#16:6",
  "#16:6",
  "#16:8",
  "#16:9",
  "#16:9",
  "#16:9",
  "#16:9",
  "#16:9",
  "#17:1",
  "#17:10",
  "#17:11",
  "#17:2",
  "#17:6",
  "#17:7",
  "#17:9",
  "#18:1",
  "#18:2",
  "#1:1",
  "#1:2",
  "#1:3",
  "#2:10",
  "#2:2",
  "#2:2",
  "#2:2",
  "#2:3",
  "#2:4",
  "#2:5",
  "#2:6",
  "#2:7",
  "#2:8",
  "#2:8",
  "#2:8",
  "#2:9",
  "#2:9",
  "#3:1",
  "#3:2",
  "#3:4",
  "#4:1",
  "#4:10",
  "#4:11",
  "#4:11",
  "#4:2",
  "#4:3",
  "#4:4",
  "#4:4",
  "#4:4",
  "#4:4",
  "#4:5",
  "#4:6",
  "#4:7",
  "#4:8",
  "#4:9",
  "#5:10",
  "#5:2",
  "#5:3",
  "#5:6",
  "#5:7",
  "#5:8",
  "#5:9",
  "#6:1",
  "#6:2",
  "#6:2",
  "#6:2",
  "#6:2",
  "#6:3",
  "#6:3",
  "#6:4",
  "#6:5",
  "#6:5",
  "#6:5",
  "#6:7",
  "#6:7",
  "#7:1",
  "#7:10",
  "#7:11",
  "#7:11",
  "#7:12",
  "#7:3",
  "#7:3",
  "#7:3",
  "#7:3",
  "#7:4",
  "#7:4",
  "#7:6",
  "#7:7",
  "#7:8",
  "#7:9",
  "#8:1",
  "#8:10",
  "#8:10",
  "#8:10",
  "#8:10",
  "#8:10",
  "#8:10",
  "#8:1",
  "#8:1",
  "#8:2",
  "#8:2",
  "#8:2",
  "#8:3",
  "#8:3",
  "#8:4",
  "#8:4",
  "#8:4",
  "#8:5",
  "#8:5",
  "#8:5",
  "#8:5",
  "#8:5",
  "#8:6",
  "#8:6",
  "#8:6",
  "#8:7",
  "#8:7",
  "#8:7",
  "#8:7",
  "#8:7",
  "#8:8",
  "#8:8",
  "#8:9",
  "#8:9",
  "#8:9",
  "#8:9",
  "#9:1",
  "#9:1",
  "#9:1",
  "#9:2",
  "",
  "#2.-time-out",
  "#2.-time-out",
  "#2.-time-out",
  "#3.-team-time-out",
  "#3.-team-time-out",
  "#3.-team-time-out",
  "#4.-passive-play",
  "#4.-passive-play",
  "#4.-passive-play",
  "#4.-passive-play",
  "#4.-passive-play",
  "#6.-definition-of-clear-chance-of-scoring",
  "#6.-definition-of-clear-chance-of-scoring",
  "#6.-definition-of-clear-chance-of-scoring",
  "#6.-definition-of-clear-chance-of-scoring",
  "#7.-intervention-by-the-timekeeper-or-a-delegate",
  "#7.-intervention-by-the-timekeeper-or-a-delegate",
  "#7.-intervention-by-the-timekeeper-or-a-delegate",
  "#7.-intervention-by-the-timekeeper-or-a-delegate",
  "#8.-injured-player",
  "#8.-injured-player",
  "#guidelines-and-interpretations",
  "#not-respecting-the-distance",
  "#disqualification-during-the-last-30-seconds",
  "#hand-signals",
  "#guidelines-and-interpretations",
  "",
  "#substitution-area-regulations",
  "#substitution-area-regulations",
  "#substitution-area-regulations",
]
`;

exports[`mapRuleToAnchor should map all available rules to the correct anchor 2`] = `
Array [
  "#10:1",
  "#10:2",
  "#10:3",
  "#10:4",
  "#11:1",
  "#11:4",
  "#11:5",
  "#12:1",
  "#12:1",
  "#12:2",
  "#13:1",
  "#13:1",
  "#13:1",
  "#13:2",
  "#13:3",
  "#13:4",
  "#13:4",
  "#13:4",
  "#13:5",
  "#13:6",
  "#13:6",
  "#13:7",
  "#13:8",
  "#14:1",
  "#14:10",
  "#14:1",
  "#14:1",
  "#14:1",
  "#14:2",
  "#14:3",
  "#14:4",
  "#14:5",
  "#14:6",
  "#14:7",
  "#14:8",
  "#14:9",
  "#15:1",
  "#15:2",
  "#15:3",
  "#15:4",
  "#15:5",
  "#15:5",
  "#15:5",
  "#15:6",
  "#15:7",
  "#15:7",
  "#15:7",
  "#15:8",
  "#15:9",
  "#16:1",
  "#16:10",
  "#16:11",
  "#16:11",
  "#16:1",
  "#16:1",
  "#16:3",
  "#16:3",
  "#16:3",
  "#16:3",
  "#16:3",
  "#16:3",
  "#16:3",
  "#16:3",
  "#16:4",
  "#16:5",
  "#16:6",
  "#16:6",
  "#16:6",
  "#16:6",
  "#16:6",
  "#16:6",
  "#16:8",
  "#16:9",
  "#16:9",
  "#16:9",
  "#16:9",
  "#16:9",
  "#17:1",
  "#17:10",
  "#17:11",
  "#17:2",
  "#17:6",
  "#17:7",
  "#17:9",
  "#18:1",
  "#18:2",
  "#1:1",
  "#1:2",
  "#1:3",
  "#2:10",
  "#2:2",
  "#2:2",
  "#2:2",
  "#2:3",
  "#2:4",
  "#2:5",
  "#2:6",
  "#2:7",
  "#2:8",
  "#2:8",
  "#2:8",
  "#2:9",
  "#2:9",
  "#3:1",
  "#3:2",
  "#3:4",
  "#4:1",
  "#4:10",
  "#4:11",
  "#4:11",
  "#4:2",
  "#4:3",
  "#4:4",
  "#4:4",
  "#4:4",
  "#4:4",
  "#4:5",
  "#4:6",
  "#4:7",
  "#4:8",
  "#4:9",
  "#5:10",
  "#5:2",
  "#5:3",
  "#5:6",
  "#5:7",
  "#5:8",
  "#5:9",
  "#6:1",
  "#6:2",
  "#6:2",
  "#6:2",
  "#6:2",
  "#6:3",
  "#6:3",
  "#6:4",
  "#6:5",
  "#6:5",
  "#6:5",
  "#6:7",
  "#6:7",
  "#7:1",
  "#7:10",
  "#7:11",
  "#7:11",
  "#7:12",
  "#7:3",
  "#7:3",
  "#7:3",
  "#7:3",
  "#7:4",
  "#7:4",
  "#7:6",
  "#7:7",
  "#7:8",
  "#7:9",
  "#8:1",
  "#8:10",
  "#8:10",
  "#8:10",
  "#8:10",
  "#8:10",
  "#8:10",
  "#8:1",
  "#8:1",
  "#8:2",
  "#8:2",
  "#8:2",
  "#8:3",
  "#8:3",
  "#8:4",
  "#8:4",
  "#8:4",
  "#8:5",
  "#8:5",
  "#8:5",
  "#8:5",
  "#8:5",
  "#8:6",
  "#8:6",
  "#8:6",
  "#8:7",
  "#8:7",
  "#8:7",
  "#8:7",
  "#8:7",
  "#8:8",
  "#8:8",
  "#8:9",
  "#8:9",
  "#8:9",
  "#8:9",
  "#9:1",
  "#9:1",
  "#9:1",
  "#9:2",
  "",
  "#2.-time-out",
  "#2.-time-out",
  "#2.-time-out",
  "#3.-team-time-out",
  "#3.-team-time-out",
  "#3.-team-time-out",
  "#4.-passives-spiel",
  "#4.-passives-spiel",
  "#4.-passives-spiel",
  "#4.-passives-spiel",
  "#4.-passives-spiel",
  "#6.-definition-einer-klaren-torgelegenheit",
  "#6.-definition-einer-klaren-torgelegenheit",
  "#6.-definition-einer-klaren-torgelegenheit",
  "#6.-definition-einer-klaren-torgelegenheit",
  "#7.-eingreifen-durch-den-zeitnehmer",
  "#7.-eingreifen-durch-den-zeitnehmer",
  "#7.-eingreifen-durch-den-zeitnehmer",
  "#7.-eingreifen-durch-den-zeitnehmer",
  "#8.-verletzter-spieler",
  "#8.-verletzter-spieler",
  "#guidelines-und-interpretationen",
  "#abstandsvergehen",
  "#disqualifikation-in-den-letzten-30-sekunden",
  "#handzeichen",
  "#guidelines-und-interpretationen",
  "",
  "#auswechselraum-reglement",
  "#auswechselraum-reglement",
  "#auswechselraum-reglement",
]
`;

exports[`mapRuleToAnchor should map all available rules to the correct anchor 3`] = `
Array [
  "#10:1",
  "#10:2",
  "#10:3",
  "#10:4",
  "#11:1",
  "#11:4",
  "#11:5",
  "#12:1",
  "#12:1",
  "#12:2",
  "#13:1",
  "#13:1",
  "#13:1",
  "#13:2",
  "#13:3",
  "#13:4",
  "#13:4",
  "#13:4",
  "#13:5",
  "#13:6",
  "#13:6",
  "#13:7",
  "#13:8",
  "#14:1",
  "#14:10",
  "#14:1",
  "#14:1",
  "#14:1",
  "#14:2",
  "#14:3",
  "#14:4",
  "#14:5",
  "#14:6",
  "#14:7",
  "#14:8",
  "#14:9",
  "#15:1",
  "#15:2",
  "#15:3",
  "#15:4",
  "#15:5",
  "#15:5",
  "#15:5",
  "#15:6",
  "#15:7",
  "#15:7",
  "#15:7",
  "#15:8",
  "#15:9",
  "#16:1",
  "#16:10",
  "#16:11",
  "#16:11",
  "#16:1",
  "#16:1",
  "#16:3",
  "#16:3",
  "#16:3",
  "#16:3",
  "#16:3",
  "#16:3",
  "#16:3",
  "#16:3",
  "#16:4",
  "#16:5",
  "#16:6",
  "#16:6",
  "#16:6",
  "#16:6",
  "#16:6",
  "#16:6",
  "#16:8",
  "#16:9",
  "#16:9",
  "#16:9",
  "#16:9",
  "#16:9",
  "#17:1",
  "#17:10",
  "#17:11",
  "#17:2",
  "#17:6",
  "#17:7",
  "#17:9",
  "#18:1",
  "#18:2",
  "#1:1",
  "#1:2",
  "#1:3",
  "#2:10",
  "#2:2",
  "#2:2",
  "#2:2",
  "#2:3",
  "#2:4",
  "#2:5",
  "#2:6",
  "#2:7",
  "#2:8",
  "#2:8",
  "#2:8",
  "#2:9",
  "#2:9",
  "#3:1",
  "#3:2",
  "#3:4",
  "#4:1",
  "#4:10",
  "#4:11",
  "#4:11",
  "#4:2",
  "#4:3",
  "#4:4",
  "#4:4",
  "#4:4",
  "#4:4",
  "#4:5",
  "#4:6",
  "#4:7",
  "#4:8",
  "#4:9",
  "#5:10",
  "#5:2",
  "#5:3",
  "#5:6",
  "#5:7",
  "#5:8",
  "#5:9",
  "#6:1",
  "#6:2",
  "#6:2",
  "#6:2",
  "#6:2",
  "#6:3",
  "#6:3",
  "#6:4",
  "#6:5",
  "#6:5",
  "#6:5",
  "#6:7",
  "#6:7",
  "#7:1",
  "#7:10",
  "#7:11",
  "#7:11",
  "#7:12",
  "#7:3",
  "#7:3",
  "#7:3",
  "#7:3",
  "#7:4",
  "#7:4",
  "#7:6",
  "#7:7",
  "#7:8",
  "#7:9",
  "#8:1",
  "#8:10",
  "#8:10",
  "#8:10",
  "#8:10",
  "#8:10",
  "#8:10",
  "#8:1",
  "#8:1",
  "#8:2",
  "#8:2",
  "#8:2",
  "#8:3",
  "#8:3",
  "#8:4",
  "#8:4",
  "#8:4",
  "#8:5",
  "#8:5",
  "#8:5",
  "#8:5",
  "#8:5",
  "#8:6",
  "#8:6",
  "#8:6",
  "#8:7",
  "#8:7",
  "#8:7",
  "#8:7",
  "#8:7",
  "#8:8",
  "#8:8",
  "#8:9",
  "#8:9",
  "#8:9",
  "#8:9",
  "#9:1",
  "#9:1",
  "#9:1",
  "#9:2",
  "",
  "#2.-tiempo-muerto",
  "#2.-tiempo-muerto",
  "#2.-tiempo-muerto",
  "#3.-tiempo-muerto-de-equipo",
  "#3.-tiempo-muerto-de-equipo",
  "#3.-tiempo-muerto-de-equipo",
  "#4.-juego-pasivo",
  "#4.-juego-pasivo",
  "#4.-juego-pasivo",
  "#4.-juego-pasivo",
  "#4.-juego-pasivo",
  "#6.--definición-de-clara-ocasión-de-gol",
  "#6.--definición-de-clara-ocasión-de-gol",
  "#6.--definición-de-clara-ocasión-de-gol",
  "#6.--definición-de-clara-ocasión-de-gol",
  "#7.-intervención/interrupción-por-parte-del-cronometrador-o-del-delegado",
  "#7.-intervención/interrupción-por-parte-del-cronometrador-o-del-delegado",
  "#7.-intervención/interrupción-por-parte-del-cronometrador-o-del-delegado",
  "#7.-intervención/interrupción-por-parte-del-cronometrador-o-del-delegado",
  "#8.-jugador-lesionado",
  "#8.-jugador-lesionado",
  "#directrices-e-interpretaciones",
  "#no-respetar-la-distancia",
  "#descalificación-durante-los-últimos-30-segundos-de-juego",
  "#gestoforma",
  "#directrices-e-interpretaciones",
  "",
  "#reglamento-relativo-a-la-zona-de-cambios",
  "#reglamento-relativo-a-la-zona-de-cambios",
  "#reglamento-relativo-a-la-zona-de-cambios",
]
`;
